proposals/route-blinding.md

```diff
@@ -0,0 +1,242 @@
+blinded hop, starting with very low values. While the fee/cltv is below the real fee of the first
+hop, the sender will get an error from `N(0)`. Once the fee/cltv proposed actually satisfies the
+first hop's requirements, the error will come from another node `N(i)` inside the blinded path.
+
+The sender can then unblind channels one-by-one by discovering their real fees/cltv and matching
+those to existing channels in the graph.
+
+To mitigate this, when nodes along the blinded path are offered an invalid HTLC, they should:
+
+* return a dummy error encrypted with a throw-away key: the sender will receive an error she can't
+  decrypt and doesn't know which node generated it
+* hold the HTLC for a random amount of time before sending the error (otherwise the sender can
+  still use timing to guess what node errored out)
+
+Even with such mitigations the sender can discover the real fees/cltv of one of the blinded
+channels. To do so she uses the correct fees and cltv for all but one channel, and for that target
+channel she tries fees/cltv until the payment succeeds. Once the payment succeeds she knows the
+approximate fees/cltv of the target channel (but since the payment succeeded, she can't continue
+probing).
+
+Maybe nodes along the blinded path could use slightly different fees/cltv than what they publicly
+advertize? Or the recipient could add some fuzzing to it to blind them more?
+
+Are those mitigations enough? Or can a clever attacker still work around them?
+
+## Tips and Tricks
+
+### Dummy hops
+
+The sender knows an upper bound on the distance between the recipient and `N(0)`. If the recipient
+is close to `N(0)`, this might not be ideal. In such cases, the recipient may add any number of
+dummy hops at the beginning of the blinded route by using `N(j) = N(r)`. The sender will not be
+able to distinguish those from normal blinded hops.
+
+Note that the recipient needs to fully validate each dummy hop to detect tampering.
+
+### Wallets and unannounced channels
+
+Route blinding is particularly useful for wallets that are connected to nodes via unannounced
+channels. Such wallets would use a single blinded hop, which effectively hides their `node_id`
+and `scid` from the sender. It obviously reveals to the blinded node that the next node is the
+final recipient, but a wallet that's not online all the time with a stable IP will never be able
+to hide that information from the nodes it connects to anyway (even with rendezvous).
+
+### Blinded trampoline route
+
+Route blinding can also be used with trampoline very easily. Instead of encrypting the
+`outgoing_channel_id`, we simply need to encrypt the `outgoing_node_id`.
+
+Each trampoline node can then decrypt the `node_id` of the next node and compute `E(i)` for the
+next trampoline node. That `E(i)` can then be sent in the outer onion payload instead of using the
+message's TLV extensions, which is even cleaner.
+
+## FAQ
```

Comment:
To make a good comparison of different protocols, we should start from future use-cases we want to cover (multi-hops SLS locks, offers, cancellable payments), lay their round-trips, decide if we want them in-band, what high-level privacy properties do we expect, how do they solve fees/cltv leaks, etc ?

Like having a detailed list of issues to solve and evaluate each proposal with regard to them ?

Reply:
It would be good to have a check-list indeed. I don't think we've got it all figured out, so it's clearly a work in progress, but it will definitely be helpful.